coroutines-extra/src/main/scala/org/coroutines/extra/AsyncAwait.scala

```diff
@@ -66,6 +66,54 @@ object AsyncAwait {
     p.future
   }
 
+  /** Thrown by `NoYieldsValidator` if a value is yielded inside an `async` block.
+   *
+   *  @param position  Where a yield occurred inside an async block.
+   *  @param message   A description of the error.
+   */
+  class YieldInsideAsyncException[P](val position: P, val message: String)
+    extends RuntimeException
+
+  def validate[C <: Context](c: C)(body: c.Tree) {
+    import c.universe._
+
+    /** Ensures that no values are yielded inside the async block.
+     *
+     *  It is similar to and shares functionality with
+     *  [[org.coroutines.AstCanonicalization.NestedContextValidator]].
+     *
+     *  @param typer  Holds the typings for the body of the coroutine. Can be generated
+     *                using `org.coroutines.common.ByTreeTyper`.
+     */
+    class NoYieldsValidator extends Traverser {
+      // return type is the lub of the function return type and yield argument types
+      def isCoroutinesPkg(q: Tree) = q match {
+        case q"org.coroutines.`package`" => true
+        case q"coroutines.this.`package`" => true
+        case t => false
+      }
+
+      override def traverse(tree: Tree): Unit = tree match {
+        case q"$qual.yieldval[$_]($_)" if isCoroutinesPkg(qual) =>
+          throw new YieldInsideAsyncException[c.universe.Position](tree.pos,
```

Comment:
Definitely. This was a holdover from an initial implementation. I have a [new PR](https://github.com/storm-enroute/coroutines/pull/34) that addresses this.